File: book/petrol-pump/chapter2/section10/AccumulatePulses.java
package chapter2.section10;

import pump.*;
import sodium.*;

public class AccumulatePulses implements Pump
{
    public Outputs create(Inputs inputs)
    {
        Behavior<UpDown> nozzle1 = inputs.eNozzle1.hold(UpDown.DOWN);
        Behavior<UpDown> nozzle2 = inputs.eNozzle2.hold(UpDown.DOWN);
        Behavior<UpDown> nozzle3 = inputs.eNozzle3.hold(UpDown.DOWN);
        Behavior<UpDown> anyNozzle = Behavior.lift((n1, n2, n3) ->
            n1 == UpDown.UP || n2 == UpDown.UP || n3 == UpDown.UP
                    ? UpDown.UP : UpDown.DOWN,
            nozzle1, nozzle2, nozzle3);

        Behavior<Double> litersDelivered =
                accumulate(inputs.eFuelPulses, inputs.calibration);

        return new Outputs()
            .setDelivery(anyNozzle.map(
                    u -> u == UpDown.UP ? Delivery.FAST1
                                        : Delivery.OFF))
            .setSaleQuantityLCD(litersDelivered.map(
                    q -> Formatters.formatSaleQuantity(q)));
    }

    private static Behavior<Double> accumulate(
            Event<Integer> ePulses, Behavior<Double> calibration)
    {
        BehaviorLoop<Integer> total = new BehaviorLoop<>();
        total.loop(ePulses.snapshot(total,
            (pulses_, total_) -> pulses_ + total_).hold(0));

        return Behavior.lift(
            (total_, calibration_) -> total_ * calibration_,
            total, calibration);
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter2/Section13/AccumulatePulses.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter2.Section13
{
public class AccumulatePulses : IPump
{
    public Outputs Create(Inputs inputs) {
        Behavior<UpDown> nozzle1 = inputs.ENozzle1.Hold(UpDown.DOWN);
        Behavior<UpDown> nozzle2 = inputs.ENozzle2.Hold(UpDown.DOWN);
        Behavior<UpDown> nozzle3 = inputs.ENozzle3.Hold(UpDown.DOWN);
        Behavior<UpDown> anyNozzle = Behavior<UpDown>.Lift((n1, n2, n3) =>
            n1 == UpDown.UP || n2 == UpDown.UP || n3 == UpDown.UP ? UpDown.UP : UpDown.DOWN, 
            nozzle1, nozzle2, nozzle3);
        Behavior<Double> litersDelivered = Accumulate(inputs.EFuelPulses, inputs.Calibration);
        return new Outputs()
            .SetDelivery(anyNozzle.Map(u => u == UpDown.UP ? Delivery.FAST1 : Delivery.OFF))
            .SetSaleQuantityLcd(litersDelivered.Map(q => Formatters.FormatSaleQuantity(q)));
    }

    private static Behavior<double> Accumulate(
            Event<int> ePulses, Behavior<double> calibration) {
        BehaviorLoop<int> total = new BehaviorLoop<int>();
        total.Loop(ePulses.Snapshot(total,(pulses_, total_) => pulses_ + total_).Hold(0));
        return Behavior<double>.Lift(
            (total_, calibration_) => total_ * calibration_,
            total, calibration);
    }
}
}
